tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,255 @@
+
+        size_t n_read = MIN(updated_requested_len, n_avail);
+        MOCK_IO_ENSURE(n_read > 0);
+
+        /* we have already verified that there is more data, so this should succeed */
+        MOCK_IO_GUARD(s2n_stuffer_read_bytes(&io_ctx->data_buffer, buf + total_read, n_read));
+        updated_requested_len -= n_read;
+        POSIX_ENSURE_GTE(updated_requested_len, 0);
+        total_read += n_read;
+
+        /* if already read the requested amount then break */
+        if (updated_requested_len == 0) {
+            break;
+        }
+
+        /* Handle if we partially read a record */
+        ssize_t remaining_len = n_avail - n_read;
+        if (remaining_len) {
+            POSIX_GUARD_RESULT(s2n_test_ktls_rewrite_prev_header_len(io_ctx, remaining_len));
+        }
+
+        /* Attempt to read multiple records (must be of the same type) */
+        if (updated_requested_len) {
+            POSIX_GUARD(s2n_stuffer_peek_char(&io_ctx->ancillary_buffer, (char *) &next_record_type));
+
+            bool no_more_data = s2n_stuffer_data_available(&io_ctx->ancillary_buffer) == 0;
+            bool next_record_different_type = next_record_type != *record_type;
+
+            if (no_more_data || next_record_different_type) {
+                break;
+            }
+        }
+    }
+
+    return total_read;
+}
+
+S2N_RESULT s2n_test_init_ktls_stuffer_io(struct s2n_connection *server, struct s2n_connection *client,
+        struct s2n_test_ktls_io_pair *io_pair)
+{
+    RESULT_ENSURE_REF(server);
+    RESULT_ENSURE_REF(client);
+    RESULT_ENSURE_REF(io_pair);
+    /* setup stuffer IO */
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&io_pair->server_in.data_buffer, 0));
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&io_pair->server_in.ancillary_buffer, 0));
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&io_pair->client_in.data_buffer, 0));
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&io_pair->client_in.ancillary_buffer, 0));
+
+    RESULT_GUARD(s2n_ktls_set_send_recv_msg_fn(s2n_test_ktls_sendmsg_stuffer_io, s2n_test_ktls_recvmsg_stuffer_io));
+    RESULT_GUARD(s2n_ktls_set_send_recv_msg_ctx(server, &io_pair->client_in, &io_pair->server_in));
+    RESULT_GUARD(s2n_ktls_set_send_recv_msg_ctx(client, &io_pair->server_in, &io_pair->client_in));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_test_validate_data(struct s2n_test_ktls_io_stuffer *ktls_io, uint8_t *expected_data, uint16_t len)
+{
+    RESULT_ENSURE_REF(ktls_io);
+    RESULT_ENSURE_REF(expected_data);
+
+    /* verify data */
+    struct s2n_stuffer *stuffer = &ktls_io->data_buffer;
+    RESULT_ENSURE_EQ(s2n_stuffer_data_available(stuffer), len);
+    uint8_t *data_ptr = s2n_stuffer_raw_read(stuffer, len);
+    RESULT_ENSURE_REF(data_ptr);
+    RESULT_ENSURE_EQ(memcmp(data_ptr, expected_data, len), 0);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_test_validate_ancillary(struct s2n_test_ktls_io_stuffer *ktls_io, uint8_t expected_record_type, uint16_t len)
```

Comment:
What exactly is the purpose of this function? It seems to be writing to a stuffer, and then checking that the stuffer has the data you wrote...But like, we already have tests that stuffers work as expected.

Reply:
Its creating `expected_ancillary_buf` from `expected_record_type` and `len`. Its then comparing that to `ktls_io->ancillary_buffer`.

The idea is to check what we wrote to `ktls_io->ancillary_buffer`

Reply:
If your goal is to check what was written to a particular buffer, you don't need to recreate a new buffer and memcmp it with the original. You can just read values off of the buffer and check that they match what you expect.

So to test the ancillary buffer, I would call s2n_stuffer_read_uint8() for the record type and then call s2n_stuffer_read_uint16() for the size of the data. And then check the output of those calls.

I also am not sure this needs to be a separate function, but I'd like to see what it looks like refactored.

Reply:
Ill make the edits. The separate fn is nice because this validation is going to be reused alot in the following PRs.